### PR TITLE
Shape info

### PR DESCRIPTION
**Pull request by mdhaber**
Add shape info to all distributions
@chrisb83 opening this up to make it easy to see the diff
@tirthasheshpatel @kai-striega if you'd like to help, we can divide and conquer, I'll start at the bottom and work my way up.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -8586,7 +8908,7 @@ class halfgennorm_gen(rv_continuous):
 
         f(x, \beta) = \frac{\beta}{\Gamma(1/\beta)} \exp(-|x|^\beta)
 
-    for :math:`x > 0`. :math:`\Gamma` is the gamma function
+    for :math:`x, \beta > 0`. :math:`\Gamma` is the gamma function
     (`scipy.special.gamma`).
 
     `gennorm` takes ``beta`` as a shape parameter for :math:`\beta`.
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py]:
```suggestion
    `halfgennorm` takes ``beta`` as a shape parameter for :math:`\beta`.
```
Not in scope, but might as well if it's easy.

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -8603,6 +8925,8 @@ class halfgennorm_gen(rv_continuous):
     %(example)s
 
     """
+    def _shape_info(self):
+        return [_ShapeInfo("beta", False, (0, np.inf), (False, False))]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8932]: This distribution seems to work without an `_argcheck` even though `beta` is constrained. Is it because the functions naturally return `nan`? Should there be an `_argcheck` anyway, for good measure?

(Or should `_argcheck` have a default implementation based on `_shape_info` that can be overridden in cases where one parameter's bounds depends on the bounds of another?)

Not really part of this PR, but it is a natural question to ask while reviewing.

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -7840,6 +8139,11 @@ class truncnorm_gen(rv_continuous):
     def _argcheck(self, a, b):
         return a < b
 
+    def _shape_info(self):
+        ia = _ShapeInfo("a", False, (-np.inf, np.inf), (True, False))
+        ib = _ShapeInfo("b", False, (-np.inf, np.inf), (False, True))
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8147]: Yup, `stats.truncnorm.pdf(x = 0, a=-np.inf, b=np.inf)` seems to work just fine.

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -7428,8 +7727,8 @@ class truncexpon_gen(rv_continuous):
     %(example)s
 
     """
-    def _argcheck(self, b):
-        return b > 0
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:7421]: Since I'm working my way bottom to top, I'm seeing this after posting a question about missing `_argcheck`. I guess there is an inherited version that requires shape parameters to be positive?
You could have provided a default `_shape_info` in `rv_continuous`, but I decided that I like that it's explicit. It will make it easier for contributors to see that it is something to consider adding.
(Oh... should probably add this to the guide for adding a new distribution.)

File: scipy/stats/_continuous_distns.py [hunk 5]
```diff
@@ -7361,6 +7657,9 @@ def _rvs(self, c, size=None, random_state=None):
     def _argcheck(self, c):
         return (c >= 0) & (c <= 1)
 
+    def _shape_info(self):
+        return [_ShapeInfo("c", False, (0, 1.0), (True, True))]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:7664]: Would be nice if docs mentioned bounds on `c` explicitly, but we can leave such things to a cleanup effort if you prefer.

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -6727,6 +6997,9 @@ class rayleigh_gen(rv_continuous):
     """
     _support_mask = rv_continuous._open_support_mask
 
+    def _shape_info(self):
+        return []
+
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:7005]: OK, I checked from the bottom up to here.
One thought is to add a check that tests that `_shape_info` is defined (and consistent with `_argcheck`?) for all distributions.

**Comment by mdhaber** [comment]
Thanks @chrisb83. Not sure if it's needed, but just a reminder not to force-push. (Some people don't mind it, but I don't like erasing history mid-review. Besides, we're almost certainly going to squash-merge anyway!)

File: scipy/stats/_continuous_distns.py [hunk 7]
```diff
@@ -6407,6 +6659,9 @@ def _argcheck(self, skew):
         # for all skew args.
         return np.ones(np.shape(skew), dtype=bool)
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py]:
```suggestion
        return np.isfinite(skew)
```
As you did below?
Don't mean to scope creep, but since you changed it elsewhere, seems reasonable to apply here.

File: scipy/stats/_continuous_distns.py [hunk 8]
```diff
@@ -6012,8 +6253,8 @@ class t_gen(rv_continuous):
     %(example)s
 
     """
-    def _argcheck(self, df):
-        return df > 0
+    def _shape_info(self):
+        return [_ShapeInfo("df", False, (0, np.inf), (False, False))]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:6261]: Finished here before taking another break.

File: scipy/stats/_continuous_distns.py [hunk 9]
```diff
@@ -5851,6 +6081,11 @@ class ncx2_gen(rv_continuous):
     def _argcheck(self, df, nc):
         return (df > 0) & (nc >= 0)
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:6082]: Noticed that it says nc > 0 in the documentation, but we can fix that another time if you prefer.

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (chrisb83) [reply scipy/stats/_continuous_distns.py:7421]: correct, by default, the shape parameters are assumed to be positive

File: scipy/stats/_continuous_distns.py (hunk 7 above)
Comment (chrisb83) [reply scipy/stats/_continuous_distns.py]: I am still confused by the parametrization of this distribution. I assume skew=inf does not make sense, let me double check

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (tirthasheshpatel) [reply scipy/stats/_continuous_distns.py:8932]: The default `_argcheck` of `rv_continuous` masks outputs with nan for negative shape parameters. So, `halfgennorm` works without a `_argcheck` and doesn't require one too.

See this: https://github.com/scipy/scipy/blob/4d66d813205fc7280633469b3e2a7276b4dfe72e/scipy/stats/_distn_infrastructure.py#L959-L969

**Comment by tirthasheshpatel** [comment]
> if you'd like to help, we can divide and conquer, I'll start at the bottom and work my way up.

I will start from the top :smile: !!

File: scipy/stats/_continuous_distns.py (hunk 7 above)
Comment (chrisb83) [reply scipy/stats/_continuous_distns.py]: inf leads to nans, so I changed it as you proposed

File: scipy/stats/tests/test_fit.py [hunk 10]
```diff
@@ -335,9 +339,12 @@ def test_basic_fit(self, dist_name):
...
-            res = stats.fit(dist, data, bounds[:-1], optimizer=self.opt)
+            bounds = bounds[:-1]
         if getattr(dist, 'pdf', False):
             data = dist.rvs(*ref, size=N, random_state=rng)
+
+        with npt.suppress_warnings() as sup:
+            sup.filter(RuntimeWarning, "overflow encountered")
             res = stats.fit(dist, data, bounds, optimizer=self.opt)
 
         assert_allclose(res.params, ref, **self.tols)
```
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:350]: This resolves a lot of the false positives.
```suggestion
        cond1 = np.allclose(res.params, ref, **self.tols)
        nllf1 = dist.nnlf(res.params, data)
        nllf2 = dist.nnlf(ref, data)
        cond2 = nllf1 < nllf2
        cond3 = np.allclose(nllf1, nllf2, rtol=1e-2)
        assert cond1 or cond2 or cond3
```

File: scipy/stats/_levy_stable/__init__.py [hunk 11]
```diff
@@ -791,6 +791,11 @@ class levy_stable_gen(rv_continuous):
     def _argcheck(self, alpha, beta):
         return (alpha > 0) & (alpha <= 2) & (beta <= 1) & (beta >= -1)
 
+    def _shape_info(self):
+        ialpha = _ShapeInfo("alpha", False, (0, 2), (False, True))
+        ibeta = _ShapeInfo("beta", False, (-1, 1), (True, True))
+        return [ialpha, ibeta]
+
```
Comment (mdhaber) [inline scipy/stats/_levy_stable/__init__.py:798]: weird - it's not picking these up. 
```
FAILED ..\..\scipy\stats\tests\test_fit.py::TestFit::test_basic_fit[levy_stable] - NameError: name '_ShapeInfo' is not defined
```

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:8932]: Thanks! I ended up discovering that after seeing a few of these.

**Comment by mdhaber** [comment]
Oops I'll merge upstream main into my main to clean this up

File: scipy/stats/_continuous_distns.py [hunk 12]
```diff
@@ -4351,6 +4528,9 @@ class invweibull_gen(rv_continuous):
     """
     _support_mask = rv_continuous._open_support_mask
 
+    def _shape_info(self):
+        return [_ShapeInfo("c", False, (-np.inf, np.inf), (False, False))]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py]:
```suggestion
        return [_ShapeInfo("c", False, (0, np.inf), (False, False))]
```

File: scipy/stats/_continuous_distns.py [hunk 13]
```diff
@@ -3965,6 +4132,11 @@ class geninvgauss_gen(rv_continuous):
     def _argcheck(self, p, b):
         return (p == p) & (b > 0)
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4147]: image
Probably could be clarified that only b > 0.

File: scipy/stats/_continuous_distns.py [hunk 14]
```diff
@@ -3713,8 +3868,7 @@ class gausshyper_gen(rv_continuous):
 
         f(x, a, b, c, z) = C x^{a-1} (1-x)^{b-1} (1+zx)^{-c}
 
-    for :math:`0 \le x \le 1`, :math:`a > 0`, :math:`b > 0`, :math:`z > -1`,
-    and :math:`C = \frac{1}{B(a, b) F[2, 1](c, a; a+b; -z)}`.
+    for :math:`0 \le x \le 1`, :math:`a,b > 0`, :math:`c` a real number, :math:`z > -1`, and :math:`C = \frac{1}{B(a, b) F[2, 1](c, a; a+b; -z)}`.
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py]: PEP8 line length

File: scipy/stats/_continuous_distns.py [hunk 15]
```diff
@@ -3190,6 +3318,12 @@ def _argcheck(self, p, a, b):
         return (np.logical_and(np.abs(b) < a, p >= 0)
                 | np.logical_and(np.abs(b) <= a, p < 0))
 
+    def _shape_info(self):
+        ip = _ShapeInfo("p", False, (-np.inf, np.inf), (False, False))
+        ia = _ShapeInfo("a", False, (0, np.inf), (True, False))
+        ib = _ShapeInfo("b", False, (-np.inf, np.inf), (False, False))
+        return [ip, ia, ib]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:3338]: This one is more complicated than most. LGTM.

File: scipy/stats/_continuous_distns.py [hunk 16]
```diff
@@ -2592,7 +2706,10 @@ class genextreme_gen(rv_continuous):
 
     """
     def _argcheck(self, c):
-        return np.where(abs(c) == np.inf, 0, 1)
+        return np.isfinite(c)
+
+    def _shape_info(self):
+        return [_ShapeInfo("c", False, (-np.inf, np.inf), (False, False))]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:2725]: The documentation doesn't define the distribution for c < 0.

File: scipy/stats/_continuous_distns.py [hunk 17]
```diff
@@ -2209,6 +2303,12 @@ class truncweibull_min_gen(rv_continuous):
     def _argcheck(self, c, a, b):
         return (a >= 0.) & (b > a) & (c > 0.)
 
+    def _shape_info(self):
+        ic = _ShapeInfo("c", False, (0, np.inf), (False, False))
+        ia = _ShapeInfo("a", False, (0, np.inf), (True, False))
+        ib = _ShapeInfo("b", False, (0, np.inf), (False, False))
+        return [ic, ia, ib]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:2316]: This distribution doesn't appear in the rendered documentation!

File: scipy/stats/_continuous_distns.py [hunk 18]
```diff
@@ -1965,6 +2049,11 @@ class f_gen(rv_continuous):
     %(example)s
 
     """
+    def _shape_info(self):
+        idfn = _ShapeInfo("dfn", False, (0, np.inf), (False, False))
+        idfd = _ShapeInfo("dfd", False, (0, np.inf), (False, False))
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:2060]: Domain of these is not documented.

File: scipy/stats/_continuous_distns.py [hunk 19]
```diff
@@ -158,6 +161,9 @@ class kstwo_gen(rv_continuous):
     %(example)s
 
     """
+    def _shape_info(self):
+        return [_ShapeInfo("n", True, (1, np.inf), (True, False))]
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:171]: Shape parameter is not documented explicitly.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:171]: And it would be nice to add`_argcheck` here and above while we're at it.

**Review by mdhaber — COMMENTED** [review]
This looks great. I've finished working my way up the list and almost everything was correct. I can't guarantee that I found all the things that should be changed, but there were a few. I know documentation and `_argcheck`s shouldn't be in scope, but maybe it would be more convenient to just fix these little things in this PR while we're at it rather than making another issue/PR.
Please open this against the main repo when you're ready!

**Comment by mdhaber** [comment]
One more PR to your branch to consider @chrisb83. I added tests for the three distributions that failed the default test. I don't think they're particularly problematic, the default test is just unlucky.

If you don't want to add the doc fixes I mentioned above, let me know. I can create a PR on the main repo just for that if you'll take a quick look/merge. But if you want to add them here, I'll review and merge this PR when it's opened against the main repo.

File: scipy/stats/_continuous_distns.py (hunk 17 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:2316]: It was just recently merged. By me. I forgot that it's new. It's in the [dev documentation](http://scipy.github.io/devdocs/reference/generated/scipy.stats.truncweibull_min.html#scipy.stats.truncweibull_min).

We weren't super clear that a and b need to be > 0.

File: scipy/stats/_continuous_distns.py (hunk 16 above)
Comment (chrisb83) [reply scipy/stats/_continuous_distns.py:2725]: this has not been fixed yet, I need to read https://en.wikipedia.org/wiki/Generalized_extreme_value_distribution

**Comment by chrisb83** [comment]
> One more PR to your branch to consider @chrisb83. I added tests for the three distributions that failed the default test. I don't think they're particularly problematic, the default test is just unlucky.

Thanks, I merged it.

I also addressed all your comments except the one on `genextreme`

**Comment by mdhaber** [comment]
OK, other than `genextreme` this looks good to open against the main repo.

File: scipy/stats/_continuous_distns.py (hunk 16 above)
Comment (chrisb83) [reply scipy/stats/_continuous_distns.py:2725]: I could not fully reconcile the SciPy parametrization and the one on Wikipedia. By checking the code, I understand that the pdf is the same for c < 0 and c > 0, only the domain differs ((1/c, inf) vs (-inf, 1/c))

**Comment by chrisb83** [comment]
> OK, other than genextreme this looks good to open against the main repo.

I will create a PR tomorrow

**Comment by mdhaber** [comment]
Opened against main at gh-15791.
